### DISCUSSION TITLE
Alias for path param does not work when hypens are used in it.

### DISCUSSION THREAD

**Discussion by msbhogal** [fastapi/fastapi#8371] [Questions] [answered]
### First Check

- [X] I added a very descriptive title to this issue.
- [X] I used the GitHub search to find a similar issue and didn't find it.
- [X] I searched the FastAPI documentation, with the integrated search.
- [X] I already searched in Google "How to X in FastAPI" and didn't find any information.
- [X] I already read and followed all the tutorial in the docs and didn't find an answer.
- [X] I already checked if it is not related to FastAPI but to [Pydantic](https://github.com/samuelcolvin/pydantic).
- [X] I already checked if it is not related to FastAPI but to [Swagger UI](https://github.com/swagger-api/swagger-ui).
- [X] I already checked if it is not related to FastAPI but to [ReDoc](https://github.com/Redocly/redoc).

### Commit to Help

- [X] I commit to help with one of those options 👆

### Example Code

```python
from fastapi import FastAPI, Path

app = FastAPI()

@app.get("/{my-id}/list")
def list_items(
        my_id: str = Path(alias="my-id")
):
    return {"aa": my_id}
```

### Description

On hitting 
```cmd
curl -X 'GET' \
  'http://localhost:8000/sample/list' \
  -H 'accept: application/json'
```

I get 404 error 
```json
{
  "detail": "Not Found"
}
```

### Operating System

Windows

### FastAPI Version

0.79.0

### Python Version

3.9.10

### Additional Context

This behaviour is working fine with Query param alias but not with Path param alias

**Comment by JarroVGIT** [comment]
This doesn’t make any sense to want; you are trying to create an invalid python variable name with a hyphen, which is never used outside the scope of python itself. Not like for example query parameters, where the name of the parameter is exposed to the end user in the URL. 

Can I ask why you would want to do this?

**Comment by msbhogal** [comment]
I wanted to do this so that the API documentation generated shall follow the same convention of using hyphens for both query param and path param

**Comment by JarroVGIT** [comment]
I see, that actually makes sense. Unfortunately, this is not supported currently, sorry.

**Comment by sak96** [comment]
is there any particular files to look for, if one wants to implement the feature ?

**Comment by JarroVGIT** [comment]
Well, it does actually work at my end (EDIT: only in SwaggerUI, not the actual endpoint itself!), just like it works for `Query`. Using the following endpoints:

```python
@app.get("/{my_id}/list")
def list_items(my_id: str = Path(alias="my-id")):
    return {"aa": my_id}


@app.get("/query")
def query_param(other_query: str = Query(alias="other-query")):
    return {"q": query_param}
```
The SwaggerUI displays it as this:
<img width="474" alt="image" src="https://user-images.githubusercontent.com/13659033/185417099-d9492d7a-b8ae-4b4c-b40f-d7af8405d67d.png">
<img width="429" alt="image" src="https://user-images.githubusercontent.com/13659033/185417258-cbc186e0-5887-4554-8233-bc7cbfec849b.png">

As you can see, only the reference in the 'path' is not adhering to the alias, but the input parameter in the SwaggerUI is showing 'my-id' as it should. Looking at the generated JSON, you can see that FastAPI is taking the path directly as is declared. So, if you want to change this behaviour, you would have to change how FastAPI is generating the openapi.json, which happens in the `openapi` namespace. I would start at `utils.py` :)

**Comment by msbhogal** [comment]
Thanks for the suggestion. I was actually looking for something that is more consistent.

This code works, but has underscored in the path and a hyphen in the parameter description.
```python
@app.get("/{my_id}/list")
def list_items(my_id: str = Path(alias="my-id")):
    return {"aa": my_id}
```
<img width="262" alt="image" src="https://user-images.githubusercontent.com/20686344/185546474-026a131d-da91-488b-aa6d-e275acd1219b.png">

The below code generates documentation in expected format, but the API does not work.
```python
@app.get("/{my-id}/list")
def list_items(my_id: str = Path(alias="my-id")):
    return {"aa": my_id}
```
<img width="248" alt="image" src="https://user-images.githubusercontent.com/20686344/185546869-bcea65c9-1703-4905-a9ae-a7f91753a416.png">

**Comment by JarroVGIT** [comment]
I have to amend my comment, my apologies. Using `alias` parameter will break the endpoint, as it will always return with:

```json
{
  "detail": [
    {
      "loc": [
        "path",
        "my-id"
      ],
      "msg": "field required",
      "type": "value_error.missing"
    }
  ]
}
```
FastAPI will trip up as it will look for `{my-id}` rather than `{my_id}`. If you only want the path name to be changed in SwaggerUI, you could [overwrite the generated](https://fastapi.tiangolo.com/advanced/extending-openapi/) openapi schema with something of your own creation. Note that you not only have to change the `path` name, but also the `parameters`, which makes it look quite complicated. I haven't optimised this at all, it just works as an example:

```python
from copy import deepcopy
from fastapi import FastAPI, Path, Query
from fastapi.openapi.utils import get_openapi
import re

app = FastAPI()


def custom_openapi():
    if app.openapi_schema:
        return app.openapi_schema

    openapi_schema = get_openapi(
        title=app.title,
        version=app.version,
        description=app.description,
        routes=app.routes,
    )
    schema_for_iteration = deepcopy(openapi_schema)
    openapi_schema["paths"] = {}

    for path, path_definition in schema_for_iteration["paths"].items():
        list_of_path_params_with_underscores = re.findall("{([a-zA-Z_]*)}", path)
        if list_of_path_params_with_underscores:
            # Change the 'path' key and replace _ with -
            new_path = path.replace("_", "-")
            openapi_schema["paths"][new_path] = schema_for_iteration["paths"][path]
            # For each path param with underscores, changes the parameter name
            for path_param in list_of_path_params_with_underscores:
                new_path_param = path_param.replace("_", "-")
                for http_verb, http_verb_definition in path_definition.items():
                    for index, param in enumerate(http_verb_definition["parameters"]):
                        if param["name"] == path_param:
                            openapi_schema["paths"][new_path][http_verb]["parameters"][
                                index
                            ]["name"] = new_path_param
        else:
            # If there are no path params with underscores, take the original definition
            openapi_schema["paths"][path] = schema_for_iteration["paths"][path]

    return openapi_schema


app.openapi = custom_openapi


@app.get("/{my_id}/list")
def list_items(my_id: str = Path()):
    return {"id": my_id}


@app.get("/query")
def query_param(other_query: str = Query(alias="other-query")):
    return {"q": query_param}


if __name__ == "__main__":
    import uvicorn

    uvicorn.run(app, host="0.0.0.0", port=8000)
```
This will create the following swaggerUI:

<img width="541" alt="image" src="https://user-images.githubusercontent.com/13659033/185562504-3ef51488-0df4-4563-afd5-dd9262354795.png">

**Comment by jrobbins-LiveData** [comment]
This route, based on Closed Issue https://github.com/tiangolo/fastapi/issues/131#issuecomment-1177187775, works, and shows how one can alias a path parameter, as long as the alias lacks a `hyphen`.


```
@router.get("/api/users/{id}/profile/")
def user_profile(user_id: int = Path(None, alias="id")):
    return {
        "id": user_id,
        "username": "johndoe"
    }
```


This route, using "user-id" instead of "id" as the alias, does not work, and returns `404 Not Found`:

```
@router.get("/api/users1/{user-id}/profile/")
def user_profile(user_id: int = Path(None, alias="user-id")):
    return {
        "id": user_id,
        "username": "johndoe"
    }
```

@tiangolo any advice on where in the code the mapping from a `Path(alias...` happens?  I would like to help fix this.

**Comment by OH-ELYADATA** [comment]
@tiangolo any idea on why it isn't working when using hyphens in path parameter aliases? is this planned to be fixed?

**Comment by RedLine89** [comment]
@tiangolo also wondering if this is going to be addressed any time soon? or if there is a workaround for this?

**Comment by ghost** [comment]
This  definitely needs to be addressed for as the best user’s experience to use dash in the end point, rather than underscore.

**Comment by sharonyogev** [comment, marked as answer]
I think it might be related to starlette, see here in the compile_path function, the regex does not allow hyphen in the variable name:
```PARAM_REGEX = re.compile("{([a-zA-Z_][a-zA-Z0-9_]*)(:[a-zA-Z_][a-zA-Z0-9_]*)?}")```
reference: https://github.com/encode/starlette/blob/0.26.1/starlette/routing.py#L108

**Comment by goharShoukat** [comment]
I stumbled over this myself. This is still erroring out. Query params work though when using aliases with hyphens
